alexandria/core/tests/test_filters.py

```diff
@@ -169,6 +171,75 @@ def test_tag_filter(
     assert received_ids == expected_ids
 
 
+def test_tag_synonym_filter(
+    db, document_factory, tag_factory, admin_client, tag_synonym_group_factory
+):
+    url = reverse("document-list")
+    ngroups = 2
+    groups = {}
+    groupsize = 3
+    tagged_group = {}
+    for i in range(1, ngroups + 1):
+        group = f"synonyms{i}"
+        groups[group] = None
+        tagged_group[group] = [f"group{i}_syn{n}" for n in range(1, groupsize + 1)]
+    ungrouped = [f"word{i}" for i in range(1, groupsize + 1)]
+
+    def create_tagged_docs(
+        synonyms: list, grouped: bool = True
+    ) -> Optional[TagSynonymGroup]:
+        try:
+            assert len(synonyms) > 0
+        except (TypeError, AssertionError) as e:
+            raise e
+
+        if grouped:
+            tag_group = tag_synonym_group_factory(
+                tags=[tag_factory(slug=word) for word in synonyms]
+            )
+            for n in range(1, len(synonyms) + 1):
+                if not tag_group.tags:
+                    return
+                perm = list(combinations(tag_group.tags.all(), n))
+                for tag_set in perm:
+                    document_factory(tags=tag_set)
+            return tag_group
+        else:
+            for color in synonyms:
+                document_factory(tags=[tag_factory(slug=color)])
+
+    for key, value in groups.items():
+        groups[key] = create_tagged_docs(tagged_group[key])
+
+    create_tagged_docs(ungrouped, False)
+
+    key = list(groups.keys())[0]
+    tag_sets_base = [
+        doc.tags.all().values_list("slug", flat=True)
+        for doc in Document.objects.filter(tags__tag_synonym_group=groups[key])
+    ]
+
+    tag_group_synonyms1_tagged_documents = Document.objects.filter(
+        tags__tag_synonym_group=groups[key]
+    )
+
+    for tags in tag_sets_base:
+        resp = admin_client.get(url, {"filter[tags]": tags})
+        assert resp.status_code == HTTP_200_OK
+        result = resp.json()
+        assert [str(doc.id) for doc in tag_group_synonyms1_tagged_documents] == [
+            obj["id"] for obj in result["data"]
+        ]
+
+    for tag in ungrouped:
+        resp = admin_client.get(url, {"filter[tags]": [tag]})
+        assert resp.status_code == HTTP_200_OK
+        result = resp.json()
+        assert [str(doc.id) for doc in Tag.objects.get(slug=tag).documents.all()] == [
+            obj["id"] for obj in result["data"]
+        ]
```

Comment:
This is potentially flaky - if the ordering of the returned documents changes, the filter would still be correct, but the test would fail. We can mitigate this by doing the following:

```suggestion
        assert set([str(doc.id) for doc in Tag.objects.get(slug=tag).documents.all()] )== set([
            obj["id"] for obj in result["data"]
        ])
```

There's the same mistake in the block above on line 230